File: user-interfaces/sdk/typescript/file-system/examples/basic-usage.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * File System SDK - Basic Usage Example
 *
 * This example demonstrates:
 * 1. Connecting to the blockchain and provider
 * 2. Creating a drive
 * 3. Creating directories
 * 4. Uploading files
 * 5. Downloading files by path
 * 6. Listing directory contents
 * 7. Deleting files
 * 8. Checking drive index root
 *
 * Prerequisites:
 *   - Parachain running at ws://127.0.0.1:2222
 *   - Provider running at http://127.0.0.1:3333
 *   - npm install && npm run papi:generate
 *
 * Usage:
 *   npx tsx examples/basic-usage.ts
 */

import { FileSystemClient } from "../src/index.js";

const CHAIN_WS = process.env.CHAIN_WS || "ws://127.0.0.1:2222";
const PROVIDER_URL = process.env.PROVIDER_URL || "http://127.0.0.1:3333";

async function main() {
  console.log("=== File System SDK - Basic Usage ===\n");
  console.log(`Chain: ${CHAIN_WS}`);
  console.log(`Provider: ${PROVIDER_URL}\n`);

  const client = new FileSystemClient({
    chainWs: CHAIN_WS,
    providerUrl: PROVIDER_URL,
  });

  try {
    // Step 1: Connect
    console.log("Step 1: Connecting...");
    await client.connect();
    await client.setSigner("//Alice");
    console.log(`  Connected as: ${client.getAddress()}\n`);

    // Step 2: Create a drive
    console.log("Step 2: Creating drive...");
    const driveId = await client.createDrive({
      name: "My TypeScript Drive",
      capacity: 1_073_741_824n, // 1 GB (2^30 bytes)
      duration: 500, // 500 blocks
      maxPayment: 1_000_000_000_000_000n, // 1000 tokens
      minProviders: 1,
    });
    console.log(`  Drive created: ID = ${driveId}\n`);

    // Step 3: Get drive info
    console.log("Step 3: Getting drive info...");
    const drive = await client.getDrive(driveId);
    console.log(`  Name: ${drive?.name}`);
    console.log(`  Bucket ID: ${drive?.bucketId}`);
    console.log(`  Owner: ${drive?.owner}\n`);

    // Step 4: Create directories
    console.log("Step 4: Creating directories...");
    await client.createDirectory(driveId, "/documents");
    console.log("  Created /documents");
    await client.createDirectory(driveId, "/documents/notes");
    console.log("  Created /documents/notes\n");

    // Step 5: Upload files
    console.log("Step 5: Uploading files...");
    const content1 = new TextEncoder().encode("Hello from TypeScript SDK!");
    const result1 = await client.uploadFile(driveId, "/documents/hello.txt", content1);
    console.log(`  Uploaded /documents/hello.txt (CID: ${result1.cid}, ${result1.size} bytes)`);

    const content2 = new TextEncoder().encode(
      JSON.stringify({ message: "Hello JSON!", timestamp: Date.now() })
    );
    const result2 = await client.uploadFile(driveId, "/documents/notes/data.json", content2, {
      contentType: "application/json",
    });
    console.log(`  Uploaded /documents/notes/data.json (${result2.size} bytes)\n`);

    // Step 6: Download and verify
    console.log("Step 6: Downloading and verifying...");
    const downloaded = await client.downloadFile(driveId, "/documents/hello.txt");
    const downloadedText = new TextDecoder().decode(downloaded.data);
    console.log(`  Downloaded: "${downloadedText}"`);
    console.log(`  Content-Type: ${downloaded.contentType}`);
    console.log(`  Size: ${downloaded.size} bytes`);
    const matches = downloadedText === "Hello from TypeScript SDK!";
    console.log(`  Verified: ${matches ? "OK" : "MISMATCH"}\n`);

    // Step 7: List directory
    console.log("Step 7: Listing /documents...");
    const listing = await client.listDirectory(driveId, "/documents");
    console.log(`  Path: ${listing.path}`);
    console.log(`  Files: ${listing.fileCount}, Dirs: ${listing.dirCount}, Total: ${listing.totalSize} bytes`);
    for (const entry of listing.entries) {
      const type = entry.entry_type === "dir" ? "[DIR]" : "[FILE]";
      console.log(`    ${type} ${entry.name} (${entry.size} bytes)`);
    }
    console.log();

    // Step 8: Check index root
    console.log("Step 8: Getting index root...");
    const indexRoot = await client.getIndexRoot(driveId);
    console.log(`  Merkle root: ${indexRoot.metadataMerkleRoot}`);
    console.log(`  Files: ${indexRoot.fileCount}, Dirs: ${indexRoot.dirCount}`);
    console.log(`  Total size: ${indexRoot.totalSize} bytes\n`);

    // Step 9: Delete a file
    console.log("Step 9: Deleting /documents/hello.txt...");
    const deleteResult = await client.deleteFile(driveId, "/documents/hello.txt");
    console.log(`  Deleted: ${deleteResult.deleted}\n`);

    // Step 10: List drives
    console.log("Step 10: Listing all drives...");
    const drives = await client.listDrives();
    console.log(`  Found ${drives.length} drive(s):`);
    for (const d of drives) {
      console.log(`    - ID ${d.driveId}: ${d.name || "(unnamed)"}`);
    }

    console.log("\n=== Example completed successfully! ===");
  } catch (error) {
    console.error("\nError:", error);
    process.exitCode = 1;
  } finally {
    client.disconnect();
  }
}

main();

File: storage-interfaces/file-system/client/examples/basic_usage.rs
// SPDX-License-Identifier: Apache-2.0

//! Basic Usage Example for File System Client
//!
//! This example demonstrates:
//! - Negotiating signed agreement terms with a provider
//! - Creating a drive (atomically opens bucket + primary agreement)
//! - Creating directories
//! - Uploading files
//! - Listing directory contents
//! - Downloading files
//!
//! Prerequisites:
//! 1. Start the blockchain: `just start-chain`
//! 2. Start a provider node: `just start-provider`
//! 3. Provider must be registered with `accepting_primary = true` (run `just demo`
//!    once or register manually via Polkadot.js).
//!
//! Run this example:
//! ```bash
//! cargo run --example basic_usage [chain_ws] [provider_url]
//! ```

use file_system_client::FileSystemClient;
use sp_runtime::AccountId32;
use std::env;
use storage_client::{NegotiateRequest, ProviderClient};
use storage_subxt::subxt_signer::sr25519::dev as dev_signer;

const DEFAULT_CHAIN_WS: &str = "ws://127.0.0.1:2222";
const DEFAULT_PROVIDER_URL: &str = "http://127.0.0.1:3333";

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    tracing_subscriber::fmt::init();

    let args: Vec<String> = env::args().collect();
    let chain_ws = args.get(1).map(|s| s.as_str()).unwrap_or(DEFAULT_CHAIN_WS);
    let provider_url = args
        .get(2)
        .map(|s| s.as_str())
        .unwrap_or(DEFAULT_PROVIDER_URL);

    println!("🚀 File System Client - Basic Usage Example\n");
    println!("{}", "=".repeat(60));
    println!("Chain WebSocket: {chain_ws}");
    println!("Provider URL: {provider_url}");

    // === STEP 1: Create the client ===
    println!("\n📡 Step 1: Connecting to blockchain and provider...");

    let mut fs_client = FileSystemClient::new(chain_ws, provider_url)
        .await?
        .with_dev_signer("alice") // Use Alice for testing
        .await?;

    let owner: AccountId32 = dev_signer::alice().public_key().0.into();
    let provider = ProviderClient::fetch_provider_id(provider_url).await?;
    println!("  Provider account (from /info): {provider}");

    println!("✅ Connected successfully!");

    // === STEP 2: Negotiate signed terms with the provider ===
    println!("\n🤝 Step 2: Negotiating signed agreement terms...");

    let signed = ProviderClient::negotiate_terms(
        provider_url,
        &NegotiateRequest {
            owner: owner.clone(),
            max_bytes: 10_000_000_000, // 10 GB
            duration: 500,             // 500 blocks
            price_per_byte: 1,
            replica_params: None,
            bucket_id: None,
        },
    )
    .await?;

    println!(
        "✅ Provider signed terms: nonce={}, valid_until={}",
        signed.terms.nonce, signed.terms.valid_until
    );

    // === STEP 3: Create a drive ===
    println!("\n📁 Step 3: Creating a new drive...");

    let drive_id = fs_client
        .create_drive(
            Some("My Documents"),
            provider,
            signed.terms,
            signed.signature,
        )
        .await?;

    println!("✅ Drive created with ID: {drive_id}");

    // === STEP 4: Create directories ===
    println!("\n📂 Step 4: Creating directory structure...");

    let bucket_id = fs_client.get_bucket_id(drive_id).await?;
    println!("   Associated bucket: ID = {bucket_id}");

    println!("   Creating /documents...");
    fs_client
        .create_directory(drive_id, "/documents", bucket_id)
        .await?;
    println!("   ✅ Created /documents");

    println!("   Creating /documents/work...");
    fs_client
        .create_directory(drive_id, "/documents/work", bucket_id)
        .await?;
    println!("   ✅ Created /documents/work");

    println!("   Creating /photos...");
    fs_client
        .create_directory(drive_id, "/photos", bucket_id)
        .await?;
    println!("   ✅ Created /photos");

    // === STEP 5: Upload files ===
    println!("\n📝 Step 5: Uploading files...");

    let readme_content = b"# My Documents\n\nWelcome to my decentralized file system!\n\nThis is a demo of Layer 1 file system built on Scalable Web3 Storage.";
    println!(
        "   Uploading /README.md ({} bytes)...",
        readme_content.len()
    );
    fs_client
        .upload_file(drive_id, "/README.md", readme_content, bucket_id)
        .await?;
    println!("   ✅ Uploaded /README.md");

    // Upload a file in subdirectory
    let report_content =
        b"Q4 2024 Report\n\n Revenue: $1M\nGrowth: 50%\nCustomers: 1000\n\nStrong quarter!";
    println!(
        "   Uploading /documents/work/report.txt ({} bytes)...",
        report_content.len()
    );
    fs_client
        .upload_file(
            drive_id,
            "/documents/work/report.txt",
            report_content,
            bucket_id,
        )
        .await?;
    println!("   ✅ Uploaded /documents/work/report.txt");

    // Upload another file
    let notes_content = b"Meeting Notes - 2024-12-01\n\n1. Discussed Q4 results\n2. Planning for 2025\n3. New hires approved";
    println!(
        "   Uploading /documents/notes.txt ({} bytes)...",
        notes_content.len()
    );
    fs_client
        .upload_file(drive_id, "/documents/notes.txt", notes_content, bucket_id)
        .await?;
    println!("   ✅ Uploaded /documents/notes.txt");

    // === STEP 6: List directory contents ===
    println!("\n📋 Step 6: Listing directory contents...");

    // List root directory
    println!("\n   Contents of /:");
    let root_entries = fs_client.list_directory(drive_id, "/").await?;
    for entry in root_entries {
        let entry_type = if entry.is_directory() { "📁" } else { "📄" };
        println!(
            "   {} {} ({} bytes)",
            entry_type,
            entry.name_str(),
            entry.size
        );
    }

    // List /documents directory
    println!("\n   Contents of /documents:");
    let docs_entries = fs_client.list_directory(drive_id, "/documents").await?;
    for entry in docs_entries {
        let entry_type = if entry.is_directory() { "📁" } else { "📄" };
        println!(
            "   {} {} ({} bytes)",
            entry_type,
            entry.name_str(),
            entry.size
        );
    }

    // List /documents/work directory
    println!("\n   Contents of /documents/work:");
    let work_entries = fs_client
        .list_directory(drive_id, "/documents/work")
        .await?;
    for entry in work_entries {
        let entry_type = if entry.is_directory() { "📁" } else { "📄" };
        println!(
            "   {} {} ({} bytes)",
            entry_type,
            entry.name_str(),
            entry.size
        );
    }

    // === STEP 7: Download and verify files ===
    println!("\n⬇️  Step 7: Downloading and verifying files...");

    // Download README.md
    println!("\n   Downloading /README.md...");
    let downloaded_readme = fs_client.download_file(drive_id, "/README.md").await?;
    println!("   ✅ Downloaded {} bytes", downloaded_readme.len());

    // Verify content
    if downloaded_readme == readme_content {
        println!("   ✅ Content verified!");
        println!(
            "   Content preview: {}",
            String::from_utf8_lossy(&downloaded_readme[..50])
        );
    } else {
        println!("   ❌ Content mismatch!");
    }

    // Download report
    println!("\n   Downloading /documents/work/report.txt...");
    let downloaded_report = fs_client
        .download_file(drive_id, "/documents/work/report.txt")
        .await?;
    println!("   ✅ Downloaded {} bytes", downloaded_report.len());

    if downloaded_report == report_content {
        println!("   ✅ Content verified!");
        let report_text = String::from_utf8_lossy(&downloaded_report);
        println!(
            "   Content:\n{}",
            report_text
                .lines()
                .take(3)
                .collect::<Vec<_>>()
                .join("\n   ")
        );
    } else {
        println!("   ❌ Content mismatch!");
    }

    // === Summary ===
    println!("\n{}", "=".repeat(60));
    println!("\n🎉 Example completed successfully!");
    println!("\n📊 Summary:");
    println!("   ✅ Created drive: {drive_id}");
    println!("   ✅ Created 3 directories");
    println!("   ✅ Uploaded 3 files");
    println!("   ✅ Listed directory contents");
    println!("   ✅ Downloaded and verified files");
    println!("\n💡 Next steps:");
    println!("   - Try clearing the drive: clear_drive()");
    println!("   - Try deleting the drive: delete_drive()");
    println!("   - Explore more file operations");
    println!("   - Check the on-chain state via polkadot.js");

    Ok(())
}
